source/extensions/filters/common/ext_authz/check_request_utils.cc

```diff
@@ -44,6 +44,39 @@ NotHeaderKeyMatcher::NotHeaderKeyMatcher(std::vector<Matchers::StringMatcherPtr>
 
 bool NotHeaderKeyMatcher::matches(absl::string_view key) const { return !matcher_.matches(key); }
 
+// Convenience function for either adding a new header or appending value to existing header.
+void addOrAppendHeader(envoy::config::core::v3::HeaderMap& mutable_header_map,
+                       absl::string_view key, absl::string_view value) {
+  auto* headers = mutable_header_map.mutable_headers();
+  for (auto& header : *headers) {
+    if (header.key() == key) {
+      // Merge duplicate headers.
```

Comment:
There shouldn't be any merge here. Every header should just be appended to the end of the array. No need to check if it exists already; it's intended that there can be duplicates.